// hardhat-gas-reporter/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OPTIMISM_GAS_ORACLE_ABI_PARTIAL = exports.OPTIMISM_GAS_ORACLE_ADDRESS = exports.UNICODE_TRIANGLE = exports.UNICODE_CIRCLE = exports.OPTIMISM_ECOTONE_BLOB_BASE_FEE_SCALAR = exports.OPTIMISM_ECOTONE_BASE_FEE_SCALAR = exports.OPTIMISM_BEDROCK_DYNAMIC_OVERHEAD = exports.OPTIMISM_BEDROCK_FIXED_OVERHEAD = exports.EVM_BASE_TX_COST = exports.TOOLCHAIN_FOUNDRY = exports.TOOLCHAIN_HARDHAT = exports.DEFAULT_ARBITRUM_HARDFORK = exports.DEFAULT_OPTIMISM_HARDFORK = exports.DEFAULT_COINMARKET_BASE_URL = exports.DEFAULT_API_KEY_ARGS = exports.DEFAULT_GAS_PRICE_API_ARGS = exports.DEFAULT_GET_BLOCK_API_ARGS = exports.DEFAULT_GAS_PRICE_PRECISION = exports.DEFAULT_JSON_OUTPUT_FILE = exports.DEFAULT_CURRENCY_DISPLAY_PRECISION = exports.DEFAULT_CURRENCY = exports.TABLE_NAME_MARKDOWN = exports.TABLE_NAME_TERMINAL = exports.TABLE_NAME_LEGACY = void 0;
exports.TABLE_NAME_LEGACY = "legacy";
exports.TABLE_NAME_TERMINAL = "terminal";
exports.TABLE_NAME_MARKDOWN = "markdown";
exports.DEFAULT_CURRENCY = "USD";
exports.DEFAULT_CURRENCY_DISPLAY_PRECISION = 2;
exports.DEFAULT_JSON_OUTPUT_FILE = "./gasReporterOutput.json";
exports.DEFAULT_GAS_PRICE_PRECISION = 5;
exports.DEFAULT_GET_BLOCK_API_ARGS = "action=eth_getBlockByNumber&tag=latest&boolean=false";
exports.DEFAULT_GAS_PRICE_API_ARGS = "action=eth_gasPrice";
exports.DEFAULT_API_KEY_ARGS = "&apikey=";
exports.DEFAULT_COINMARKET_BASE_URL = "https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/";
exports.DEFAULT_OPTIMISM_HARDFORK = "ecotone";
exports.DEFAULT_ARBITRUM_HARDFORK = "arbOS11";
exports.TOOLCHAIN_HARDHAT = "hardhat";
exports.TOOLCHAIN_FOUNDRY = "foundry";
// EVM
exports.EVM_BASE_TX_COST = 21000;
// Source:
// https://docs.optimism.io/stack/transactions/fees#bedrock
exports.OPTIMISM_BEDROCK_FIXED_OVERHEAD = 188;
exports.OPTIMISM_BEDROCK_DYNAMIC_OVERHEAD = 0.684;
// These params are configured by node operators and may vary
// Values are suggested default values from:
// https://docs.optimism.io/builders/chain-operators/management/blobs
exports.OPTIMISM_ECOTONE_BASE_FEE_SCALAR = 1368;
exports.OPTIMISM_ECOTONE_BLOB_BASE_FEE_SCALAR = 810949;
exports.UNICODE_CIRCLE = "◯";
exports.UNICODE_TRIANGLE = "△";
exports.OPTIMISM_GAS_ORACLE_ADDRESS = "0xb528d11cc114e026f138fe568744c6d45ce6da7a";
exports.OPTIMISM_GAS_ORACLE_ABI_PARTIAL = [
    {
        constant: true,
        inputs: [],
        name: "blobBaseFee",
        outputs: [
            {
                name: "",
                type: "uint256",
            },
        ],
        payable: false,
        stateMutability: "view",
        type: "function",
    },
    {
        constant: true,
        inputs: [],
        name: "baseFeeScalar",
        outputs: [
            {
                name: "",
                type: "uint32",
            },
        ],
        payable: false,
        stateMutability: "view",
        type: "function",
    },
    {
        constant: true,
        inputs: [],
        name: "blobBaseFeeScalar",
        outputs: [
            {
                name: "",
                type: "uint32",
            },
        ],
        payable: false,
        stateMutability: "view",
        type: "function",
    }
];
